File: lin-compiler/lin_compiler/ScriptWrite.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Linq;

namespace LIN
{
    static class ScriptWrite
    {
        static public void WriteSource(Script s, string Filename, int indentSpaces = 2)
        {
            Program.PrintLine("[write] writing decompiled file...");

            using (var File = new System.IO.StreamWriter(Filename, false, Encoding.UTF8))
            {
                var outputBuilder = new StringBuilder();
                int setOptionIndent = 0;
                int checkObjectIndent = 0;
                int checkCharacterIndent = 0;

                // First pass: Mark opcodes to skip (those that follow Text entries)
                bool[] skipEntry = new bool[s.ScriptData.Count];
                for (int i = 0; i < s.ScriptData.Count; i++)
                {
                    ScriptEntry e = s.ScriptData[i];

                    // Check if this is a Text opcode
                    if (e.Opcode == 0x02 && e.Text != null)
                    {
                        bool hasCLTTags = System.Text.RegularExpressions.Regex.IsMatch(e.Text, @"<CLT\s+\d+>|<CLT>");

                        // Mark the TextStyle before this Text (if present and has CLT tags)
                        if (hasCLTTags && i > 0 && s.ScriptData[i - 1].Opcode == 0x03)
                        {
                            skipEntry[i - 1] = true;
                        }

                        // Mark all following WaitFrame and WaitInput opcodes after ANY Text entry
                        // Also mark TextStyle opcodes if the text has CLT tags
                        int lookAhead = i + 1;
                        while (lookAhead < s.ScriptData.Count)
                        {
                            byte nextOpcode = s.ScriptData[lookAhead].Opcode;

                            // Always skip WaitFrame (0x3B) and WaitInput (0x3A) after Text
                            // Also skip TextStyle (0x03) if Text has CLT tags
                            if (nextOpcode == 0x3B || nextOpcode == 0x3A || (hasCLTTags && nextOpcode == 0x03))
                            {
                                skipEntry[lookAhead] = true;
                                lookAhead++;
                                // Stop after WaitInput
                                if (nextOpcode == 0x3A)
                                    break;
                            }
                            else
                            {
                                break;
                            }
                        }
                    }
                }

                // Second pass: Write entries
                for (int i = 0; i < s.ScriptData.Count; i++)
                {
                    ScriptEntry e = s.ScriptData[i];

                    // Skip entries marked in first pass
                    if (skipEntry[i])
                    {
                        continue;
                    }

                    // Skip Type opcode (0x00) - it's inferred from presence of Text opcodes
                    if (e.Opcode == 0x00)
                    {
                        continue;
                    }

                    BaseOpcode opcode = OpcodeDictionary.GetOpcodeDefinition(e.Opcode);

                    // Check if this is SetOption to adjust indentation
                    if (opcode != null && opcode.Name == "SetOption")
                    {
                        // SetOption(255) closes the current level
                        if (e.Args.Length > 0 && e.Args[0] == 255)
                        {
                            setOptionIndent = 0;
                        }
                        else
                        {
                            // Any other SetOption resets to base level (not nested)
                            setOptionIndent = 0;
                        }
                    }

                    // Check if this is CheckObject to adjust indentation
                    if (opcode != null && opcode.Name == "CheckObject")
                    {
                        // CheckObject(255) closes the CheckObject indentation
                        if (e.Args.Length > 0 && e.Args[0] == 255)
                        {
                            checkObjectIndent = 0;
                        }
                        // Any other CheckObject resets to base level (not nested)
                        else if (e.Args.Length > 0)
                        {
                            checkObjectIndent = 0;
                        }
                    }

                    // Check if this is CheckCharacter to adjust indentation
                    if (opcode != null && opcode.Name == "CheckCharacter")
                    {
                        // CheckCharacter(255) closes the CheckCharacter indentation
                        if (e.Args.Length > 0 && e.Args[0] == 255)
                        {
                            checkCharacterIndent = 0;
                        }
                        // Any other CheckCharacter resets to base level (not nested)
                        else if (e.Args.Length > 0)
                        {
                            checkCharacterIndent = 0;
                        }
                    }

                    // Write indentation (additive: SetOption + CheckObject + CheckCharacter)
                    int totalIndent = setOptionIndent + checkObjectIndent + checkCharacterIndent;
                    string indent = new string(' ', totalIndent * indentSpaces);
                    outputBuilder.Append(indent);

                    if (opcode == null)
                    {
                        // Unknown opcode - write as hex with raw bytes
                        outputBuilder.Append($"0x{e.Opcode:X2}(");
                        if (e.Args.Length > 0)
                        {
                            outputBuilder.AppendJoin(", ", e.Args.Select(b => b.ToString()));
                        }
                        outputBuilder.AppendLine(")");
                    }
                    else
                    {
                        opcode.WriteSource(outputBuilder, s, e);
                    }

                    // After writing SetOption, increase indent for content (except for 255)
                    if (opcode != null && opcode.Name == "SetOption")
                    {
                        if (e.Args.Length > 0 && e.Args[0] != 255)
                        {
                            setOptionIndent = 1;
                        }
                    }

                    // After writing CheckObject, increase indent for content (except for 255)
                    if (opcode != null && opcode.Name == "CheckObject")
                    {
                        if (e.Args.Length > 0 && e.Args[0] != 255)
                        {
                            checkObjectIndent = 1;
                        }
                    }

                    // After writing CheckCharacter, increase indent for content (except for 255)
                    if (opcode != null && opcode.Name == "CheckCharacter")
                    {
                        if (e.Args.Length > 0 && e.Args[0] != 255)
                        {
                            checkCharacterIndent = 1;
                        }
                    }
                }

                File.Write(outputBuilder.ToString());
            }

            Program.PrintLine("[write] done.");
        }


        static public void WriteCompiled(Script s, string Filename)
        {
            Program.PrintLine("[write] writing compiled file...");

            // Infer script type from presence of Text opcodes
            bool hasTextOpcodes = s.ScriptData.Any(e => e.Opcode == 0x02);
            s.Type = hasTextOpcodes ? ScriptType.Text : ScriptType.Textless;

            List<byte> File =
            [
                // Header
                .. BitConverter.GetBytes((Int32)s.Type),
                .. BitConverter.GetBytes(s.Type == ScriptType.Text ? 16 : 12),
            ];
            switch (s.Type)
            {
                case ScriptType.Textless:
                    File.AddRange(BitConverter.GetBytes(s.FileSize));
                    break;
                case ScriptType.Text:
                    File.AddRange(BitConverter.GetBytes(s.TextBlockPos));
                    File.AddRange(BitConverter.GetBytes(s.FileSize));
                    break;
                default: throw new Exception("[write] error: unknown script type.");
            }

            Dictionary<int, string> TextData = [];
            if (s.Type == ScriptType.Text)
            {
                s.TextEntries = 0;
                // First pass: assign text IDs to all text opcodes
                foreach (ScriptEntry e in s.ScriptData)
                {
                    if (e.Opcode == 0x02)
                    {
                        while (TextData.ContainsKey(s.TextEntries)) s.TextEntries++;
                        TextData.Add(s.TextEntries, e.Text);

                        BaseOpcode opcode = OpcodeDictionary.GetOpcodeDefinition(e.Opcode);
                        opcode.PrepareForCompilation(s, e);
                    }
                }

                s.TextEntries = TextData.Keys.Count > 0
                    ? Math.Max(s.TextEntries, TextData.Keys.Max() + 1)
                    : 0;
            }

            // Create and insert Type opcode at the beginning
            ScriptEntry typeEntry = new ScriptEntry()
            {
                Opcode = 0x00,
                Args = new byte[2]
            };
            BaseOpcode typeOpcode = OpcodeDictionary.GetOpcodeDefinition(0x00);
            typeOpcode.PrepareForCompilation(s, typeEntry);

            // Write Type opcode first
            File.Add(0x70);
            File.Add(typeEntry.Opcode);
            File.AddRange(typeEntry.Args);

            // Write remaining opcodes (skip any existing Type opcodes from source)
            foreach (ScriptEntry e in s.ScriptData)
            {
                // Skip Type opcodes - we already wrote one at the beginning
                if (e.Opcode == 0x00)
                {
                    continue;
                }

                File.Add(0x70);
                File.Add(e.Opcode);
                File.AddRange(e.Args);
            }

            while (File.Count % 4 != 0) File.Add(0x00);

            s.TextBlockPos = File.Count;
            for (int i = 0; i < 4; i++) File[0x08 + i] = BitConverter.GetBytes(s.TextBlockPos)[i];

            if (s.Type == ScriptType.Textless)
            {
                s.FileSize = s.TextBlockPos;
            }
            else if (s.Type == ScriptType.Text)
            {
                File.AddRange(BitConverter.GetBytes(s.TextEntries));
                int[] StartPoints = new int[s.TextEntries];
                int Total = 8 + s.TextEntries * 4;

                for (int i = 0; i < s.TextEntries; i++)
                {
                    // Add the BOM
                    string Text = "\uFFFE";
                    if (TextData.ContainsKey(i)) Text = TextData[i];
                    if (Text.Length == 0 || Text[Text.Length - 1] != '\0') Text += '\0';

                    byte[] ByteText = Encoding.Unicode.GetBytes(Text);
                    if (ByteText[0] != 0xFF || ByteText[1] != 0xFE)
                    {
                        byte[] Temp = new byte[ByteText.Length + 2];
                        Temp[0] = 0xFF;
                        Temp[1] = 0xFE;
                        ByteText.CopyTo(Temp, 2);
                        ByteText = Temp;
                    }

                    StartPoints[i] = Total;
                    Total += ByteText.Length;
                }

                foreach (int sp in StartPoints)
                {
                    File.AddRange(BitConverter.GetBytes(sp));
                }
                File.AddRange(BitConverter.GetBytes(Total));

                for (int i = 0; i < s.TextEntries; i++)
                {
                    string Text = "\uFFFE";
                    if (TextData.ContainsKey(i)) Text = TextData[i];
                    if (Text.Length == 0 || Text[Text.Length - 1] != '\0') Text += '\0';

                    byte[] ByteText = Encoding.Unicode.GetBytes(Text);
                    if (ByteText[0] != 0xFF || ByteText[1] != 0xFE)
                    {
                        byte[] Temp = new byte[ByteText.Length + 2];
                        Temp[0] = 0xFF;
                        Temp[1] = 0xFE;
                        ByteText.CopyTo(Temp, 2);
                        ByteText = Temp;
                    }

                    File.AddRange(ByteText);
                }

                while (File.Count % 4 != 0) File.Add(0x00);
                s.FileSize = File.Count;
                for (int i = 0; i < 4; i++) File[0x0C + i] = BitConverter.GetBytes(s.FileSize)[i];
            }
            System.IO.File.WriteAllBytes(Filename, [.. File]);
            Program.PrintLine("[write] done.");
        }
    }
}

File: lin-compiler/src/scriptWrite.ts
import { writeFile } from "node:fs/promises";
import { OP_TEXT, OP_TEXT_STYLE, OP_TYPE, OP_WAIT_FRAME, OP_WAIT_INPUT } from "./opcodes/ids.js";
import { getOpcodeDefinition } from "./opcodes/opcodeDictionary.js";
import { printLine, toHexOpcode } from "./options.js";
import { SourceBuilder } from "./output.js";
import { type Script, type ScriptEntry, ScriptType } from "./script.js";

const UTF8_BOM = "\uFEFF";
const HAS_CLT = /<CLT\s+\d+>|<CLT>/;

/** Opcodes whose 255 argument closes a block, and whose other values open one. */
const INDENTING_OPCODES = ["SetOption", "CheckObject", "CheckCharacter"] as const;

export function writeSourceText(s: Script, indentSpaces = 2): string {
  const output = new SourceBuilder();

  // Indent depth contributed by each block-opening opcode; they sum additively
  const indentLevels = new Map<string, number>(INDENTING_OPCODES.map((name) => [name, 0]));

  // First pass: determine which Text entries can use AutoText sugar, and mark opcodes to skip
  const { useAutoText, skipEntry } = planAutoText(s);

  // Second pass: write entries
  for (let i = 0; i < s.scriptData.length; i++) {
    const e = s.scriptData[i];

    // Skip entries marked in the first pass
    if (skipEntry[i]) {
      continue;
    }

    // Skip Type opcode - it's inferred from the presence of Text opcodes
    if (e.opcode === OP_TYPE) {
      continue;
    }

    const opcode = getOpcodeDefinition(e.opcode);
    const isIndenting = opcode !== null && indentLevels.has(opcode.name ?? "");

    // A block-opening opcode always closes the previous block first
    if (isIndenting && e.args.length > 0) {
      indentLevels.set(opcode.name as string, 0);
    }

    // Write indentation (additive across all block-opening opcodes)
    const totalIndent = [...indentLevels.values()].reduce((a, b) => a + b, 0);
    output.append(" ".repeat(totalIndent * indentSpaces));

    if (opcode === null) {
      // Unknown opcode - write as hex with raw bytes
      output.append(`${toHexOpcode(e.opcode)}(`);
      if (e.args.length > 0) {
        output.appendJoin(", ", e.args.map(String));
      }
      output.appendLine(")");
    } else if (e.opcode === OP_TEXT && useAutoText[i]) {
      // Write as AutoText instead of Text
      output.append("AutoText(");
      opcode.writeSourceArgs(output, s, e);
      output.appendLine(")");
    } else {
      opcode.writeSource(output, s, e);
    }

    // After writing, increase the indent for the block's content (except for 255, which closes it)
    if (isIndenting && e.args.length > 0 && e.args[0] !== 255) {
      indentLevels.set(opcode.name as string, 1);
    }
  }

  return output.toString();
}

export async function writeSource(s: Script, filename: string, indentSpaces = 2): Promise<void> {
  printLine("[write] writing decompiled file...");
  // .NET's StreamWriter with Encoding.UTF8 emits a BOM; keep the output byte-compatible
  await writeFile(filename, UTF8_BOM + writeSourceText(s, indentSpaces), "utf8");
  printLine("[write] done.");
}

/**
 * Finds Text entries that can be collapsed into AutoText sugar: a Text followed only by
 * WaitFrame (and TextStyle, when the text carries CLT tags) and terminated by WaitInput.
 */
function planAutoText(s: Script): { useAutoText: boolean[]; skipEntry: boolean[] } {
  const useAutoText = new Array<boolean>(s.scriptData.length).fill(false);
  const skipEntry = new Array<boolean>(s.scriptData.length).fill(false);

  for (let i = 0; i < s.scriptData.length; i++) {
    const e = s.scriptData[i];

    if (e.opcode !== OP_TEXT || e.text == null) {
      continue;
    }

    const hasCLTTags = HAS_CLT.test(e.text);

    // A preceding TextStyle belongs to the CLT tags and is regenerated on compile
    const hasMatchingPrecedingTextStyle = hasCLTTags && i > 0 && s.scriptData[i - 1].opcode === OP_TEXT_STYLE;

    let waitInputIndex = -1;
    let lookAhead = i + 1;
    while (lookAhead < s.scriptData.length) {
      const nextOpcode = s.scriptData[lookAhead].opcode;

      if (nextOpcode === OP_WAIT_INPUT) {
        waitInputIndex = lookAhead;
        break;
      }
      if (nextOpcode === OP_WAIT_FRAME || (nextOpcode === OP_TEXT_STYLE && hasCLTTags)) {
        lookAhead++;
        continue;
      }
      // Non-sugar opcode found before WaitInput
      break;
    }

    if (waitInputIndex === -1) {
      continue;
    }

    useAutoText[i] = true;

    if (hasMatchingPrecedingTextStyle) {
      skipEntry[i - 1] = true;
    }

    // Mark all opcodes between Text and WaitInput (inclusive) to skip
    for (let j = i + 1; j <= waitInputIndex; j++) {
      skipEntry[j] = true;
    }
  }

  return { useAutoText, skipEntry };
}

/** Growable byte buffer with the little-endian int32 writes the LIN format uses. */
class ByteWriter {
  private bytes: number[] = [];

  get length(): number {
    return this.bytes.length;
  }

  push(...values: number[]): void {
    this.bytes.push(...values);
  }

  pushMany(values: ArrayLike<number>): void {
    for (let i = 0; i < values.length; i++) {
      this.bytes.push(values[i]);
    }
  }

  pushInt32(value: number): void {
    this.bytes.push(value & 0xff, (value >>> 8) & 0xff, (value >>> 16) & 0xff, (value >>> 24) & 0xff);
  }

  setInt32(offset: number, value: number): void {
    this.bytes[offset] = value & 0xff;
    this.bytes[offset + 1] = (value >>> 8) & 0xff;
    this.bytes[offset + 2] = (value >>> 16) & 0xff;
    this.bytes[offset + 3] = (value >>> 24) & 0xff;
  }

  padTo4(): void {
    while (this.bytes.length % 4 !== 0) {
      this.bytes.push(0x00);
    }
  }

  toBuffer(): Buffer {
    return Buffer.from(this.bytes);
  }
}

/** Encode a text entry as UTF-16LE with a leading byte-order mark and a null terminator. */
function encodeTextEntry(text: string): Buffer {
  const terminated = text.endsWith("\0") ? text : `${text}\0`;
  const encoded = Buffer.from(terminated, "utf16le");
  if (encoded[0] === 0xff && encoded[1] === 0xfe) {
    return encoded;
  }
  return Buffer.concat([Buffer.from([0xff, 0xfe]), encoded]);
}

export function writeCompiledBytes(s: Script): Buffer {
  // Infer script type from presence of Text opcodes
  const hasTextOpcodes = s.scriptData.some((e) => e.opcode === OP_TEXT);
  s.type = hasTextOpcodes ? ScriptType.Text : ScriptType.Textless;

  const file = new ByteWriter();

  // Header
  file.pushInt32(s.type);
  file.pushInt32(s.type === ScriptType.Text ? 16 : 12);
  if (s.type === ScriptType.Text) {
    file.pushInt32(s.textBlockPos);
  }
  file.pushInt32(s.fileSize);

  const textData = new Map<number, string>();
  if (s.type === ScriptType.Text) {
    s.textEntries = 0;
    // First pass: assign text IDs to all text opcodes
    for (const e of s.scriptData) {
      if (e.opcode === OP_TEXT) {
        textData.set(s.textEntries, e.text ?? "");
        getOpcodeDefinition(e.opcode)?.prepareForCompilation(s, e);
      }
    }
  }

  // Create and insert the Type opcode at the beginning
  const typeEntry: ScriptEntry = { opcode: OP_TYPE, args: [0, 0] };
  getOpcodeDefinition(OP_TYPE)?.prepareForCompilation(s, typeEntry);

  file.push(0x70, typeEntry.opcode);
  file.pushMany(typeEntry.args);

  // Write remaining opcodes (skipping any Type opcodes from the source)
  for (const e of s.scriptData) {
    if (e.opcode === OP_TYPE) {
      continue;
    }
    file.push(0x70, e.opcode);
    file.pushMany(e.args);
  }

  file.padTo4();

  s.textBlockPos = file.length;
  file.setInt32(0x08, s.textBlockPos);

  if (s.type === ScriptType.Textless) {
    s.fileSize = s.textBlockPos;
    return file.toBuffer();
  }

  file.pushInt32(s.textEntries);

  // Text block layout: an offset table (one entry per text, plus an end marker) then the strings
  const encoded: Buffer[] = [];
  const startPoints: number[] = [];
  let total = 8 + s.textEntries * 4;

  for (let i = 0; i < s.textEntries; i++) {
    const bytes = encodeTextEntry(textData.get(i) ?? "\uFFFE");
    encoded.push(bytes);
    startPoints.push(total);
    total += bytes.length;
  }

  for (const startPoint of startPoints) {
    file.pushInt32(startPoint);
  }
  file.pushInt32(total);

  for (const bytes of encoded) {
    file.pushMany(bytes);
  }

  file.padTo4();
  s.fileSize = file.length;
  file.setInt32(0x0c, s.fileSize);

  return file.toBuffer();
}

export async function writeCompiled(s: Script, filename: string): Promise<void> {
  printLine("[write] writing compiled file...");
  await writeFile(filename, writeCompiledBytes(s));
  printLine("[write] done.");
}
